network/simple-onchain-discovery/src/lib.rs

```diff
@@ -301,6 +307,88 @@ impl ValidatorSetChangeListener {
     }
 }
 
+/// Watcher for a file containing discovered peers at a given interval
+pub struct FileChangeListener {
+    network_context: Arc<NetworkContext>,
+    file_path: PathBuf,
+    update_channel: Mutex<channel::Sender<ConnectivityRequest>>,
+    time_service: TimeService,
+    interval_duration: Duration,
+    delay: Option<Pin<Box<Sleep>>>,
+}
+
+impl FileChangeListener {
+    pub fn new(
+        network_context: Arc<NetworkContext>,
+        file_path: &Path,
+        update_channel: channel::Sender<ConnectivityRequest>,
+        interval_duration: Duration,
+        time_service: TimeService,
+    ) -> Self {
+        FileChangeListener {
+            network_context,
+            file_path: file_path.to_path_buf(),
+            update_channel: Mutex::new(update_channel),
+            time_service,
+            interval_duration,
+            delay: None,
+        }
+    }
+}
+
+#[async_trait]
+impl DiscoveryChangeListenerTrait for FileChangeListener {
+    fn network_context(&self) -> &NetworkContext {
+        &self.network_context
+    }
+
+    fn discovery_source(&self) -> DiscoverySource {
+        DiscoverySource::File
+    }
+
+    fn update_channel(&self) -> &Mutex<channel::Sender<ConnectivityRequest>> {
+        &self.update_channel
+    }
+}
+
+impl<'a> Stream for FileChangeListener {
+    type Item = PeerSet;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        // Wait for delay, or add the delay for next call
+        if let Some(ref mut delay) = self.delay {
+            futures::ready!(Pin::new(delay).poll(cx));
+        }
+        self.delay = Some(Box::pin(self.time_service.sleep(self.interval_duration)));
+
+        match load_file(self.file_path.as_path()) {
+            Ok(peers) => Poll::Ready(Some(peers)),
+            Err(error) => {
+                // TODO: Do we keep trying to load the file?
```

Comment:
This might hint that you need to change the type returned from the Stream to be a Result. Eg if you end up doing a network call for discovery you'd need to have a way to indicate that you errored out in the rpc request

Reply:
I'd prefer to let the stream implementation handle this.  The only layer above this is a simple message shipper to the connectivity manager.